src/toxcore/dht/dht_friend.rs

```diff
@@ -0,0 +1,372 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Module for friend.
+*/
+
+use std::time::{Duration, Instant};
+use std::io::{Error, ErrorKind};
+
+use futures::{future, Future, stream, Stream};
+use parking_lot::RwLock;
+use std::sync::Arc;
+use std::ops::Deref;
+
+use toxcore::dht::packed_node::*;
+use toxcore::dht::kbucket::*;
+use toxcore::crypto_core::*;
+use toxcore::dht::server::*;
+use toxcore::dht::server::client::*;
+use toxcore::io_tokio::IoFuture;
+
+/// Hold friend related info.
+pub struct DhtFriend {
+    /// Public key of friend
+    pk: PublicKey,
+    /// close nodes of friend
+    close_nodes: Arc<RwLock<Bucket>>,
```

Comment:
Do you really need RwLock for all these fields? It seems to me that you can take `&mut self` in all methods.

Reply:
In the future, DhtFriend will be used in various case.
Especially close_nodes should be maintained one object even if it is cloned.
I don't know the trade-off of permitting mutable self vs. RwLock.
But &mut self may cause problem in the future.

Reply:
Ah. that's right.